### ISSUE TITLE
Exit Node: ACL and Client Visibility

### ISSUE THREAD

**Issue by Qubitium** [tailscale/tailscale#1567]
Tailscale 1.6 now has Exit Node feature which is great but it suffers the same ACL visibility issue as regular node list.

Expect:
------------
Node N is assigned Exit Node status
User A does not have access to Node N via ACL
User B does have access to Node N via ACL

Only user B should have access to Node N Exit Node via Windows/Android/iOS/Mac GUI. Should not allow User B to activate Exit Node that violates ACL. Should not even show Node N as an option for Exit Node. 

Actual
-------------
Both User A and B can select and activate the exit node but it will work work with User B.  Activating Exit Node N for User A will just bring all packets offline as if network is completely off. 

There is no visual feedback telling user A that this Exit Node is not usable. 

Unlike the Node List ACL visibility issue, Exit Node visibility/activation issue is worsened by the fact the client **activates** the node in **green color** and the user fully expects to use it.

**Comment by danderson** [comment]
Thanks for the report. You're right, we should filter exit node availability by ACLs. This depends on the netmap minimization change we're doing for node visibility, but might need a bit more work for exit nodes.

**Comment by DentonGentry** [comment]
Possibly relevant: netmap minimization deployed for all users a couple months ago.

**Comment by hadican** [comment]
Hi!

Any updates about this?

**Comment by stefanocerio** [comment]
hello, any updates on this? it's annoying that users who do not have access to exit node (because of ACL rule) still see the exit node on their apps.

**Comment by hadican** [comment]
@stefanocerio Hi, how did you manage to restrict access to exit node? I tried everything, but couldn’t manage it to work.

**Comment by shousper** [comment]
Hellooo, just ran up against this shortcoming. Would love to know if this is on a near-term roadmap 🙏🏻

**Comment by shousper** [comment]
Hi @DentonGentry, I think I might need an example? I tried what I thought you meant, but it granted users of the group access to use _all exit nodes in the tailnet_ not just certain tagged devices.

From the original issue description, I think this statement is the problem:

> Only user B should have access to Node N Exit Node

To give a complete example, we have an ACL rule as follows:

```
{
  "action": "accept",
  "users": ["group:developers"],
  "ports": ["tag:environment-staging:*"]
}
```

This grants our developers full access to all devices tagged as part of our staging environment. If I add `autogroup:internet` to the `ports`, like so:

```
  "ports": ["tag:environment-staging:*", "autogroup:internet:*"]
```

This doesn't limit them to exit nodes tagged as staging, it gives them access to ALL exit nodes in the tailnet. Is there another way to achieve this? If not, is it on a roadmap? 😃

Thanks in advance 🙏🏻

**Comment by codyro** [comment]
> This doesn't limit them to exit nodes tagged as staging, it gives them access to ALL exit nodes in the tailnet. Is there another way to achieve this? If not, is it on a roadmap? 😃

I was just looking to do this and the documentation now states the following while linking to this GitHub issue -- so I would imagine it's somewhere on their roadmap.

> Right now, there is no way to restrict use of specific exit nodes using ACLs. Subscribe to [this GitHub issue](https://github.com/tailscale/tailscale/issues/1567) for updates.

**Comment by GlitchWitch** [comment]
@danderson Can you comment on the above? [This section](https://tailscale.com/kb/1018/acls/#subnet-routers-and-exit-nodes) of the documentation links back to this issue. 

Is adding the ability to restrict use of specific exit nodes using ACLs actually on the roadmap?

**Comment by DentonGentry** [comment]
In most cases when ability to block specific exit nodes is brought up, beyond ability to block `autogroup:internet`, it is because the exit node is being used as a way to get a known IP address in order to access a remote Software as a Service provider and there is a desire to control access to the SaaS service. Many of those SaaS services, in turn, require an IP address for the client to put into their allowlist.

In these circumstances we recommend a different solution: a subnet router. One can set up a subnet router for any IP addresses, not just YOUR IP addresses. If the service provider offers their services on 1.2.3.4 and 5.6.7.8, one can set up a subnet router with `--advertise-routes=1.2.3.4/32,5.6.7.8/32` which will make traffic to that SaaS service go over the tailnet and egress via the subnet router.

One can then write ACLs to control access to `1.2.3.4/32` and `5.6.7.8/32`.

--------

ACLs controlling access to regular exit nodes is something we're likely to implement some point, but does not have a schedule yet. ACL control of access to *any* subnet router is implemented now, by allowing `autogroup:internet`

**Comment by GlitchWitch** [comment]
Thanks for the reply @DentonGentry. While that isn't our use case here, I think even in the example you provide that begins to fall apart for a lot of SaaS apps that use a WAF like CloudFlare and do not have a publicly available static set of routes of their own.

It sort of seems like a double edged sword. The fact that Tailscale does not have a way to limit which users or groups can access a specific exit node severely limits the use cases of exit nodes to small networks where robust access controls aren't important. Because the use case becomes so limited without the ability to scale, it gets deprioritized.

**Comment by GlitchWitch** [comment]
Thanks for the reply @DentonGentry. While that isn't our use case here, I think even in the example you provide that begins to fall apart for a lot of SaaS apps that use a WAF like CloudFlare and do not have a publicly available static set of routes of their own.

It sort of seems like a double edged sword. The fact that Tailscale does not have a way to limit which users or groups can access a specific exit node severely limits the use cases of exit nodes to small networks where robust access controls aren't important. Because the use case becomes so limited without the ability to scale, it gets deprioritized.

Adding to the confusion here, it seems the initial issue and the one posed by @shousper, @codyro, and myself are different.

**Comment by DentonGentry** [comment]
> I think even in the example you provide that begins to fall apart for a lot of SaaS apps that use a WAF like CloudFlare and do not have a publicly available static set of routes of their own.

This would be better handled by https://github.com/tailscale/tailscale/issues/1748, not an exit node.

I understand the feature request: exit nodes are a way to get a predictable public IP as the source address, so the instinct is to ask for enhancements to exit nodes. However I believe the product would be better served by features which let it steer the specific traffic of interest. Exit nodes redirect a firehose in order to get a glass of water.

>  The fact that Tailscale does not have a way to limit which users or groups can access a specific exit node severely limits the use cases of exit nodes to small networks where robust access controls aren't important

For sites using exit nodes for the purpose for which they are intended, the main requests concern automatically choosing the nearest exit node or helping ensure their users enable exit nodes at the right times. Requests for ACL controls or policy-based selection of exit nodes are usually attempts to solve a different problem using the exit node feature as the basis. Exit nodes are not necessarily the right solution.

If you're not comfortable describing the actual problem to be solved on a public bug, sending a description to info@tailscale.com would let us associate it with other similar feature requests.

**Comment by patbranchaud** [comment]
Restrict visibility and access between exit nodes controlled with ACLs would be very helpful. 
Example : Exit node A do not see Exit node B in Tailscale status or user interface.

Can we expect such feature in a near future?

_Apart of that little must have, Tailscale is just extraordinary. Thank you_

**Comment by DentonGentry** [comment]
> Restrict visibility and access between exit nodes controlled with ACLs would be very helpful.
> Example : Exit node A do not see Exit node B in Tailscale status or user interface.

That should already work. If no ACL allows connectivity between A and B, then the coordination server will not inform A and B of the existence of the other node and they will not show up in tailscale status or the user interface.

If they can connect, even in one direction and not the other, they will appear in each other's netmaps and be visible in tailscale status.

**Comment by GlitchWitch** [comment]
> Exit nodes redirect a firehose in order to get a glass of water.

Clearly people want and need that firehose. Your view of the use case for exit nodes still continues to be limited and does not align with one if the key ways VPNs are used. If we want specific users in our organisation to have a specific IPs when browsing the public internet then a firehose is warranted.

> exit nodes are a way to get a predictable public IP as the source address, so the instinct is to ask for enhancements to exit nodes.

You're correct. The reality here is many organisations (mine included) and users already use other products like OpenVPN to ensure predictable public IPs for ALL outbound requests. Being able to restrict which users on our network can use and see specific exit nodes would help bring Tailscale up to feature parity.

> That should already work. If no ACL allows connectivity between A and B, then the coordination server will not inform A and B of the existence of the other node and they will not show up in tailscale status or the user interface.

The documentation I linked to before still indicates the following:

> Right now, there is no way to restrict use of specific exit nodes using ACLs.

This also aligns with what support indicates...

**Comment by rweirtbg** [comment]
I agree with @GlitchWitch 100%. We really need this feature in order to have more granular access control over traffic flows.

**Comment by codyro** [comment]
This has cropped up in another organization that I'm part of that's moved to TailScale. Is this something on the roadmap or not? The hand wavy nature from TailScale regarding this issue leads me to believe it's not considered a priority even though there is a clear need for this functionality. 

What's the verdict @DentonGentry ?

**Comment by DentonGentry** [comment]
I do expect to extend ACLs to control access to specific exit nodes based on the user's identity or tags or other factors.

I expect https://github.com/tailscale/tailscale/issues/1748 to be implemented first, and this functionality to be worked on after that.

**Comment by cyn110** [comment]
And 100% agree with @GlitchWitch. Please increase the priority of this issue. The subnet router solution is not a validate solution for a lot of the use cases.

**Comment by DentonGentry** [comment]
> The subnet router solution is not a validate solution for a lot of the use cases.

Can you explain what those use cases are?

**Comment by cyn110** [comment]
> > The subnet router solution is not a validate solution for a lot of the use cases.
> 
> Can you explain what those use cases are?

When the SaaS doesn't have static IPs for one obviously. Also, the subnet router solution is simply not a elegant solution for us, so we can't use Tailscale for our use case.

@GlitchWitch already said very well. I think you should listen to your users.

**Comment by SGprooo** [comment]
> The subnet router solution is not a validate solution for a lot of the use cases.
Agreeing on the increased priority! Many SaaS services might use CDNs or WAFs that do not have static IP addresses. In these cases, it's difficult to configure subnet routers to route traffic to these dynamic IPs.
Plus, having granular access control over exit nodes is essential for netadmins to meet security, compliance, performance, and user management requirements. It's a critical feature that would significantly enhance applicability of Tailscale in many different scenarios.

**Comment by GlitchWitch** [comment]
@DentonGentry Do you have any stats for how many people click the link to this issue in the support documentation?

> Right now, there is no way to restrict use of specific exit nodes using ACLs. Subscribe to [this GitHub issue](https://github.com/tailscale/tailscale/issues/1567) for updates.
From: https://tailscale.com/kb/1018/acls/#subnet-routers-and-exit-nodes

This issue has been open for more than 2 years now, which represents the majority of your time working at Tailscale. It seems clear that users are asking for this and that your proposals don't work for everyone, perhaps we can get an ETA or a read from someone at Tailscale who is less invested in telling people they are wrong?

**Comment by GlitchWitch** [comment]
@DentonGentry Do you have any stats for how many people click the link to this issue in the support documentation?

> Right now, there is no way to restrict use of specific exit nodes using ACLs. Subscribe to [this GitHub issue](https://github.com/tailscale/tailscale/issues/1567) for updates.
From: https://tailscale.com/kb/1018/acls/#subnet-routers-and-exit-nodes

This issue has been open for more than 2 years now, which represents the majority of your time working at Tailscale. It seems clear that users are asking for this and that your proposals don't work for everyone, perhaps we can get an ETA or a read from someone at Tailscale who is less invested in telling people they are wrong?

You already saw (and minimised) a comment from a user in #1748 who switched away due for a similar reason. Have you considered the business impact of the number of users who silently have done this as well?

**Comment by diegomontoya** [comment]
"What we've got here is failure to communicate. " - Guns N Roses

I believe what tailscale staff perceive the exit-node feature to be and what the users effected by this issue is a difference on the fundamental understanding of what this feature is intended for. 

So users want a:

1. Reverse subnet-router. 
2. Instead of whitelist of routes to a router, a negative everything-but-these-routes to the routes
3. Instead of route 11.22.33.0/24 to a subnet-router, route everything that is not tailscale dst ip into this router
4. Add visible/toggle to this router so user (if acl assigned, can use this)

What I propose not "fixing" this issue but adding the above features to the sub-net router system and allow them to be exposed to UI as a toggle on/off and UI visibility restriction on user ACL. This would effectively REPLACE the exit-node feature and make the subnet router much more powerful/flexible.

**Comment by GlitchWitch** [comment]
@diegomontoya What you've described is not what myself and several other users have asked.

The exit node functionality works as intended, users simply want to have ACLs that limit which users can use which exit nodes. The current system means if you add an exit node, every single user with access to exit nodes can use _any_ exit node.

**Comment by SeSeKenny** [comment]
I will add my voice here, control over which exit nodes are seen by which users, hopefully via ACL automatically, if not another schema extension to the ACL json to define whitelists of users/groups/tags of allows on exit nodes would be fine too. Admins would love it to test and ensure experiences globally, for users would be very useful for traffic engineering. Agree the domains is a nice to have, but I'd argue we have an uncontrollable firehose now, pointing it should be second, then controlling it with a faucet tertiary.

**Comment by DentonGentry** [comment]
We've updated [#1748](https://github.com/tailscale/tailscale/issues/1748) with current status and plans, to direct certain domains to a subnet router for egress.

I do expect to address ACL controls to allow specification of users and groups allowed to use a given exit node. This will require additions to the ACL syntax, right now ACLs cannot express "allow $USER to reach $DESTINATION by way of $NODE." Current ACLs only express $USER and $DESTINATION.

The planned work on [#1748](https://github.com/tailscale/tailscale/issues/1748) needs to be finished first, and other enhancements in the plumbing for ACLs need to be done before we'd be in a position to add those capabilities to the ACL syntax.

**Comment by patbranchaud** [comment]
Thank you, Denton, for the update. We are all very excited about the upcoming feature. I'm wondering if my excitement is justified and would appreciate it if you could validate this use case:

Computer A is a server that uses computers B, C, and D as exit nodes. Although A can view B, C, and D in the Tailscale UI / Tailscale Status, I aim to achieve complete isolation among B, C, and D. Specifically, I would like to configure the ACL such that they cannot see each other in the Tailscale UI / Tailscale Status.

Employing ACLs would be practical, especially considering that I need to apply this configuration to 60 different exit nodes.

Keep up the good work. Thank you!

**Comment by gurchik** [comment]
I came across this issue after asking the support team how to limit access to specific apps and I was told it isn't yet possible. 

I run an app connector (e.g. to Salesforce) so that traffic to the app goes through a specific public IP address. I have an ACL which grants access to the `10.0.0.0/16` subnet to many users, and the app connector happens to run in this subnet. My goal is to only allow access to the app connector to a few users, but it looks like everyone with access to the subnet has access to the app connector.

Support told me this isn't yet possible. From reading the comments here, it seems like app connectors were supposed to address some of these use cases, but maybe it's not done yet? To be clear, this is impossible right now, right, or am I missing something?

**Comment by ArthurSource** [comment]
Indeed, very disappointing :(! Are there any plans to fix it soon, or maybe the Tailscale team can suggest temporary workarounds?

**Comment by james-nofrixion** [comment]
> > The subnet router solution is not a validate solution for a lot of the use cases.
> 
> Can you explain what those use cases are?

When it would be a complete shambles to explicitly define the destination routes (e.g. different environments such as dev/sandbox/prod for something like Azure/AWS PaaS services), but is very easy to identify the source routes (e.g. exit node IP).

**Comment by hazcod** [comment]
FYI, the ACL part of exit nodes can supposedly be fixed by making use of the new `grants` feature in TS ACLs: https://tailscale.com/kb/1324/acl-grants
I checked with TS support and they're explicit, any `grant` not allowed should default to a `deny`.
We're testing this feature this week for resolving the exit node ACL woes.

e.g.:
```hujson
"ipsets": {
	"ipset:any": ["0.0.0.0/0"]
},

"grants": [
{
	// only allow the devs to use internal exit nodes
	"src": [
		"group:devs@company.com",
	],
	"srcPosture": [
		"posture:protectedWindows",
		"posture:protectedMacOS",
	],
	"dst": ["autogroup:internet", "ipset:any"],
	//"via": ["*"], // ommitting means any exit node
	"ip": ["*"]
},
{
	// fallback: allow everyone to use regular VPN exit node
	"src": ["autogroup:member"],
	"dst": ["autogroup:internet", "ipset:any"],
	"via": ["tag:vpn"],
	"ip": ["*"]
},
],
```

**Comment by Qubitium** [comment]
@hazcod  Does this also fix the visibility side of the exit-nodes?

**Comment by hazcod** [comment]
> @hazcod Does this also fix the visibility side of the exit-nodes?

No, only the ACL part.

**Comment by james-nofrixion** [comment]
I can also confirm that @hazcod's solution using grants fixes the issues with both exit node access and visibility (in the current beta version).
- tag the exit node/subnet router
-  specify the exit node tag in the 'via' section of the grant and the permitted user/group in the 'src' section.

**Comment by clstokes** [comment]
Hi folks, [Route filtering with via](https://tailscale.com/kb/1378/via) is the solution to this and that page includes [examples](https://tailscale.com/kb/1378/via#examples) for exit nodes and other use cases. With `via` you can control which subnet routers, exit nodes, etc. are made available/visible to end users.

Please open a separate issue if you have additional use cases not covered by `via`.
